src/Razor/src/Microsoft.AspNetCore.Razor.Language/DefaultRazorIntermediateNodeLoweringPhase.cs

```diff
@@ -1184,6 +1184,18 @@ public override void VisitMarkupElement(MarkupElementSyntax node)
                             BuildSourceSpanFromNode(node.EndTag), node.EndTag.GetTagNameWithOptionalBang()));
                 }
 
+                if (node.StartTag != null)
+                {
+                    var startTagName = node.StartTag.GetTagNameWithOptionalBang();
+                    if (startTagName != null && startTagName.Length > 0 && !char.IsLower(startTagName, 0))
```

Comment:
Yeah I think `char.IsUpper` makes sense. For cases like, `<!Component>` we don't want to provide a warning because we already know this was not intended to be a component because of the bang.

Reply:
> For cases like, `<!Component>` we don't want to provide a warning because we already know this was not intended to be a component because of the bang.

Yup, totally agree. Should make sure how char.IsUpper handles `!` to ensure it doesn't create the diagnostic is all.

Reply:
> Each time something in your .razor source resolves as an element (not a component), check that its name starts with a lowercase character in a-z. If we find an element whose first char isn't in a-z case sensitively, emit a warning

This was in the issue description given by @rynowak. @rynowak was this wording intentional? Would we changing the behavior if we use `char.IsUpper` instead of `!char.IsLower`?

Reply:
> Should make sure how char.IsUpper handles ! to ensure it doesn't create the diagnostic is all.

It returns `false` for `!`. We're good.